### PR TITLE
Refinery: kindlyToNull and byTrying

### PR DESCRIPTION
**Pull request by nhaagen**
As a successor of https://github.com/ILIAS-eLearning/ILIAS/pull/2899,
this still
 - fixes https://mantis.ilias.de/view.php?id=28861 for trunk
but also
 - adds a kindlyToNull-Transform
 - uses it with the ByTrying-Transform
and 
- moves the ProblemHandler-callable into a trait.

### CODE REVIEW

File: src/Refinery/Factory.php [hunk 1]
```diff
@@ -175,4 +176,13 @@ public function uri() : URI\Group
     {
         return new URI\Group();
     }
+
+    /**
+     * Accepts Transformations and uses first successful one.
+     * @param Transformation[]
```
Comment (klees) [inline src/Refinery/Factory.php]:
```suggestion
     * @param Transformation[] $transformations
```

File: src/Refinery/KindlyTo/Transformation/NullTransformation.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        );
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function __invoke($from)
+    {
+        return $this->transform($from);
+    }
```
Comment (klees) [inline src/Refinery/KindlyTo/Transformation/NullTransformation.php]: Use `DeriveInvokeFromTransform` instead.

File: src/Refinery/KindlyTo/Transformation/NullTransformation.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    public function accepts($value) : bool
+    {
+        return is_null($value) || (is_string($value) && trim($value) === '');
+    }
```
Comment (klees) [inline src/Refinery/KindlyTo/Transformation/NullTransformation.php]: We don't need that IMO. This is something only `Constraint`s need but not defined on the `Transformation`-interface.

File: src/Refinery/To/Transformation/StringTransformation.php [hunk 4]
```diff
@@ -38,4 +38,9 @@ public function __invoke($from)
     {
         return $this->transform($from);
     }
+
+    public function accepts($value) : bool
+    {
+        return is_string($value);
+    }
```
Comment (klees) [inline src/Refinery/To/Transformation/StringTransformation.php]: I don't think we need that. Am I wrong?

File: src/Refinery/ByTrying.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        };
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function __invoke($from)
+    {
+        return $this->transform($from);
+    }
```
Comment (klees) [inline src/Refinery/ByTrying.php:53]: Use `DeriveInvokeFromTransform`.

File: tests/Refinery/ByTrying/ByTryingTransformTest.php [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            'empty string3' => ['  ', null],
+            'null' => [null, null],
+            'string' => ['str', self::ERROR],
+            'int' => [1, 1],
+            'negative int' => [-1, -1],
+            'zero' => [0, 0],
+            'array' => [[], self::ERROR],
+            'bool (false)' => [false, self::ERROR],
+            'bool (true)' => [true, self::ERROR]
+        ];
```
Comment (klees) [inline tests/Refinery/ByTrying/ByTryingTransformTest.php:40]: Pleasuse PHPUnits dataproviders here (and also in the other test in this case).

File: src/Refinery/ByTrying.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * @inheritdoc
+     */
+    public function transform($from)
+    {
+        foreach ($this->transformations as $transformation) {
+            if ($transformation->accepts($from)) {
```
Comment (klees) [inline src/Refinery/ByTrying.php]: `accepts` is no method on the `Transformation`-interface, so we cannot use it here. I think we could just `applyTo` instead and use the first result that is `ok`.

File: src/Refinery/ProblemBuilder.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+namespace ILIAS\Refinery;
+
+use ILIAS\Refinery\ConstraintViolationException;
+
+trait ProblemBuilder
+{
+    /**
+     * @var string|callable
+     */
+    protected $error;
```
Comment (klees) [inline src/Refinery/ProblemBuilder.php]: Maybe we should add this as an abstract method here, to force our users to always give an error. Just having a property makes it possible to just forget to set this.

**Review by klees — CHANGES_REQUESTED** [review]
Hi @nhaagen,

this looks mostly fine, special kudos for factoring out that `ProblemBuilder`-trait. Looks useful.

I think there are some confusions regarding `Transformation` and `Constraint` that show in your code. The relation is like this: every `Constraint` is a `Transformation` but not the other way round. A `Constraint` is a special kind of `Transformation` that does not actually change values (i.e. transforms them) but rather only looks into the values and throws (or creates `Error`s) if the values do not conform to some rule. This allows to have a more specific base class for constraints where we can define more methods than we would want for generic `Transformation`s.

Thanks for working on this so carefully!!!

Best regards!

**Comment by mjansenDatabay** [comment]
@nhaagen LGTM (except the tiny changes @klees already requested). Indeed kudos for the `ProblemBuilder` trait.

File: src/Refinery/ProblemBuilder.php (hunk 8 above)
Comment (mjansenDatabay) [reply src/Refinery/ProblemBuilder.php]: Something like `abstract protected getError() : string|callable;` or `abstract protected getDefaultError() : string|callable;`  (I know these return types do not work, yet)? I would appreciate this.

**Comment by nhaagen** [comment]
Thank you both very much for your comments and hints.
I think, I got them all.

Best regards.

**Comment by klees** [comment]
Thanks @nhaagen!

**Pull request merged**